libc/string_cpy.c

```diff
@@ -0,0 +1,1382 @@
+	int i;
+
+	TEST_ASSERT_EQUAL_INT(0, big_string());
+
+
+	for (i = 0; i < sizeof(bigIntDest) / sizeof(int); i++) {
+		/*
+		 * Convert a number from uint to its representation in char and use memcpy to do a copy of the exact
+		 * element which converted uint is pointing
+		 */
+		TEST_ASSERT_EQUAL_PTR(&bigStrDest[(unsigned char)bigIntDest[i]], memccpy(bigStrDest, long_str, bigIntDest[i], BIG_NUMB));
+	}
+
+	memset(bigStrDest, 0, BIG_NUMB);
+	/* checking if stop point value is unsigned char */
+	memccpy(bigStrDest, long_str, -MAX_ASCII_SET, BIG_NUMB);
+	TEST_ASSERT_EQUAL_INT(MAX_ASCII_SET - 1, strlen(bigStrDest));
+
+#endif
+}
+
+
+TEST(string_memccpy, data_types)
+{
+#ifdef __phoenix__
+	TEST_IGNORE();
+#endif
+#ifndef __phoenix__
+
+	int num = 12345678,
+		numDest = 0;
+	float flt = 2.32252,
+		  fltDest = 0;
+	ptrdiff_t ptr = 0x2345,
+			  ptrDest = 0;
+	size_t size = SIZE_MAX,
+		   sizDest = 0;
+
+	/* Checking ability to copy numbers between places */
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(&numDest, &num, 'x', sizeof(num)));
+	TEST_ASSERT_NOT_EMPTY(&numDest);
+	TEST_ASSERT_EQUAL_INT(num, numDest);
+
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(&fltDest, &flt, 'x', sizeof(flt)));
+	TEST_ASSERT_NOT_EMPTY(&fltDest);
+	TEST_ASSERT_FLOAT_IS_DETERMINATE(fltDest);
+	TEST_ASSERT_EQUAL_FLOAT(flt, fltDest);
+
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(&ptrDest, &ptr, 'x', sizeof(ptr)));
+	TEST_ASSERT_NOT_EMPTY(&ptrDest);
+	TEST_ASSERT_EQUAL_INT(ptr, ptrDest);
+
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(&sizDest, &size, 'x', sizeof(size)));
+	TEST_ASSERT_NOT_EMPTY(&sizDest);
+	TEST_ASSERT_EQUAL_DOUBLE(size, sizDest);
+
+#endif
+}
+
+
+TEST(string_memccpy, adjacent)
+{
+#ifdef __phoenix__
+	TEST_IGNORE();
+#endif
+#ifndef __phoenix__
+
+	char testStr[] = "TEST",
+		 memStr[MAX_STR_LEN] = "\0\0\0\0\0\0\0\0\0\0TEST",
+		 expVal[MAX_STR_LEN] = "\0\0\0\0\0\0TESTTEST",
+		 testStrLen = sizeof(testStr) - 1;
+
+	/*
+	 * Copy string in his memory space right before his original place to get effect
+	 * like using strcat but in opposite direction
+	 */
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(&memStr[10 - testStrLen], &memStr[10], 'x', testStrLen));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(memStr, expVal, sizeof(memStr));
+
+	/*
+	 * Add up another iteration of string to expected value.
+	 * We must use address of any string data because strcat looking for last occurrence
+	 * of text and we filled first palaces with nulls that means its free place to put text
+	 */
+	strcat(&expVal[10], testStr);
+
+	/* Copy text in similar way how do it strcat*/
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(&memStr[10 + testStrLen], &memStr[10], 'x', testStrLen));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(memStr, expVal, sizeof(memStr));
+
+#endif
+}
+
+
+TEST(string_memccpy, one_byte)
+{
+#ifdef __phoenix__
+	TEST_IGNORE();
+#endif
+#ifndef __phoenix__
+
+	char testArray[MAX_ASCII_SET] = { 0 },
+		 input[MAX_ASCII_SET] = { 1 };
```

Comment:
It may be misleading as you had an occasion to experience, this way it is clear that the whole buffer is set to zeros:

```suggestion
		 input[MAX_ASCII_SET] = { 0 };
```